File: backend/app/rate_limit.py
"""In-process rate limiting.

Deliberately simple: one sliding window per client, plus a global concurrency
cap. Correct for a single worker, which is what the Compose setup runs. Behind
several workers this needs a shared store (Redis `INCR` + `EXPIRE`); the
interface below is what would be swapped out.
"""

from __future__ import annotations

import asyncio
import time
from collections import defaultdict, deque

from .config import settings
from .errors import ScanError


class SlidingWindowRateLimiter:
    def __init__(self, *, limit: int, window_seconds: int) -> None:
        self._limit = limit
        self._window = window_seconds
        self._hits: defaultdict[str, deque[float]] = defaultdict(deque)
        self._lock = asyncio.Lock()

    async def check(self, key: str, *, now: float | None = None) -> None:
        """Record one hit for `key`, or raise ScanError if over the limit."""
        current = time.monotonic() if now is None else now
        async with self._lock:
            hits = self._hits[key]
            cutoff = current - self._window
            while hits and hits[0] <= cutoff:
                hits.popleft()

            if len(hits) >= self._limit:
                retry_after = max(1, int(hits[0] + self._window - current) + 1)
                raise ScanError(
                    "rate_limited",
                    f"Limite de {self._limit} scans por {self._window}s atingido. "
                    f"Tente de novo em {retry_after}s.",
                    status_code=429,
                )

            hits.append(current)

    async def prune(self, *, now: float | None = None) -> None:
        """Drop windows that fully expired, so idle clients stop costing memory."""
        current = time.monotonic() if now is None else now
        async with self._lock:
            cutoff = current - self._window
            stale = [key for key, hits in self._hits.items() if not hits or hits[-1] <= cutoff]
            for key in stale:
                del self._hits[key]

    @property
    def tracked_keys(self) -> int:
        return len(self._hits)

    def retry_after_seconds(self) -> int:
        return self._window


limiter = SlidingWindowRateLimiter(
    limit=settings.rate_limit_requests,
    window_seconds=settings.rate_limit_window_seconds,
)

#: Bounds how much outbound work the process does at once, independent of who
#: asked for it. Without this, N slow targets tie up N event-loop tasks.
scan_slots = asyncio.Semaphore(settings.max_concurrent_scans)

File: backend/internal/api/ratelimit.go
package api

import (
	"sync"
	"time"
)

// SlidingWindowLimiter is deliberately simple: one sliding window per client,
// held in process memory.
//
// Correct for a single worker, which is what the Compose setup runs. It is a
// per-instance best effort on Netlify, where the platform may run several
// Lambda instances concurrently and none of them can see the others' counters
// -- there, the binding limit is the one configured at Netlify's edge, and
// this one only bounds what a single warm instance will do. Swapping in a
// shared store (Redis INCR + EXPIRE) is the other way out; this interface is
// what would be replaced.
type SlidingWindowLimiter struct {
	limit  int
	window time.Duration

	mu   sync.Mutex
	hits map[string][]time.Time
}

func NewSlidingWindowLimiter(limit int, window time.Duration) *SlidingWindowLimiter {
	return &SlidingWindowLimiter{limit: limit, window: window, hits: map[string][]time.Time{}}
}

// Allow records one hit for key. It returns false and the seconds to wait when
// the caller is over the limit.
func (l *SlidingWindowLimiter) Allow(key string, now time.Time) (bool, int) {
	l.mu.Lock()
	defer l.mu.Unlock()

	cutoff := now.Add(-l.window)
	hits := l.hits[key]
	kept := hits[:0]
	for _, hit := range hits {
		if hit.After(cutoff) {
			kept = append(kept, hit)
		}
	}

	if len(kept) >= l.limit {
		l.hits[key] = kept
		retryAfter := int(kept[0].Add(l.window).Sub(now).Seconds()) + 1
		if retryAfter < 1 {
			retryAfter = 1
		}
		return false, retryAfter
	}

	l.hits[key] = append(kept, now)
	return true, 0
}

// Prune drops windows that fully expired, so idle clients stop costing memory.
func (l *SlidingWindowLimiter) Prune(now time.Time) {
	l.mu.Lock()
	defer l.mu.Unlock()

	cutoff := now.Add(-l.window)
	for key, hits := range l.hits {
		if len(hits) == 0 || !hits[len(hits)-1].After(cutoff) {
			delete(l.hits, key)
		}
	}
}

// TrackedKeys is the number of clients currently held in memory.
func (l *SlidingWindowLimiter) TrackedKeys() int {
	l.mu.Lock()
	defer l.mu.Unlock()
	return len(l.hits)
}

// RetryAfterSeconds is the window length, the worst case a blocked caller waits.
func (l *SlidingWindowLimiter) RetryAfterSeconds() int {
	return int(l.window.Seconds())
}
